keps/sig-auth/3304-client-go-tls-config/README.md

```diff
@@ -0,0 +1,607 @@
+-->
+
+###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?
+
+<!--
+Look at the [existing SLIs/SLOs].
+
+Think about adding additional work or introducing new steps in between
+(e.g. need to do X to start a container), etc. Please describe the details.
+
+[existing SLIs/SLOs]: https://git.k8s.io/community/sig-scalability/slos/slos.md#kubernetes-slisslos
+-->
+
+###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?
+
+<!--
+Things to keep in mind include: additional in-memory state, additional
+non-trivial computations, excessive access to disks (including increased log
+volume), significant amount of data sent and/or received over network, etc.
+This through this both in small and large cases, again with respect to the
+[supported limits].
+
+[supported limits]: https://git.k8s.io/community//sig-scalability/configs-and-limits/thresholds.md
+-->
+
+### Troubleshooting
+
+<!--
+This section must be completed when targeting beta to a release.
+
+For GA, this section is required: approvers should be able to confirm the
+previous answers based on experience in the field.
+
+The Troubleshooting section currently serves the `Playbook` role. We may consider
+splitting it into a dedicated `Playbook` document (potentially with some monitoring
+details). For now, we leave it here.
+-->
+
+###### How does this feature react if the API server and/or etcd is unavailable?
+
+###### What are other known failure modes?
+
+<!--
+For each of them, fill in the following information by copying the below template:
+  - [Failure mode brief description]
+    - Detection: How can it be detected via metrics? Stated another way:
+      how can an operator troubleshoot without logging into a master or worker node?
+    - Mitigations: What can be done to stop the bleeding, especially for already
+      running user workloads?
+    - Diagnostics: What are the useful log messages and their required logging
+      levels that could help debug the issue?
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+-->
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+N/A
+
+## Alternatives
```

Comment:
you can pass your transport to client-go via https://pkg.go.dev/k8s.io/client-go/rest#RESTClientForConfigAndClient 

> func RESTClientForConfigAndClient(config *[Config](https://pkg.go.dev/k8s.io/client-go/rest#Config), 
> 
> RESTClientForConfigAndClient returns a RESTClient that satisfies the requested attributes on a client Config object. Unlike RESTClientFor, RESTClientForConfigAndClient allows to pass an http.Client that is shared between all the API Groups and Versions. Note that the http client takes precedence over the transport values configured. The http client defaults to the `http.DefaultClient` if nil.

is not a valid alternative to manage the transport externally and use it to create the clientset?

Reply:
@aojea many transports/clientsets are constructed internally and are not exposed to callers in a way that they can influence the `http.Client` that is used.

Reply:
yeah,  I've talked with @liggitt recently about that, a good example are the internal plugins that build their transport by influencing the configuration of the transport .... and this is causing problems like the recent transport cache not being reusable by the exec plugin ... 

I think to make client-go more sustainable we should stop mutating the transports internally and move towards a model where we have a REST layer and a Transport layer, and both are independently generated, we have it now with `RESTClientForConfigAndClient` ... but I miss context on these other dependencies that you mention, however, I'll be happy to work on decoupling it if there is the desire ... 

an example (is just an example) on how moving to this model solves some of the current problems is here https://github.com/kubernetes/kubernetes/pull/112176, the exec plugin is cacheable if instead of mutating the Config it consumes the transport generated from the Config and adds its own Transport characteristic on top

Reply:
We cannot remove the TLS config and authenticator cache without breaking existing users, and as I mentioned in your linked PR, we cannot mutate the transport.  I also do not believe we will be able to separate the rest config from the transport config because the rest config contains transport semantics like the TLS config details.  In general I believe that the `http.RoundTripper` makes this difficult because the request and the response are not sufficient to encapsulate all the behavior we want (i.e. TLS config, proxy config, dialer config, etc).

Reply:
> In general I believe that the http.RoundTripper makes this difficult because the request and the response are not sufficient to encapsulate all the behavior we want (i.e. TLS config, proxy config, dialer config, etc).

yeah, totally agree that is not simple and some things are impossible but I'm optimistic and I see some room for improvement :sweat_smile: